### PR TITLE
Dashboard

### PR DESCRIPTION
**Pull request by oskarprzybylski23**
## Description

**1. Built the main dashboard to hold user's widgets. The Dashboard contains components:**
- Dashboard section: wraps around dashboard content as children and takes title as a prop
-  Dashboard container: wraps around dashboard sections as direct children
-  Dashboard layout: component purpose-built for our dashboard utilizing other components as building blocks. This is the component that is then used within the pages.

**2. Added widgetContainer compoenent.**
- It is based on widgetHeader component and serves the same purpose of holding widgets. Changes are visual.

**3. AddWidgets function within the Dashboard component:**
-  takes GoalsData (fetched from API) and filters based on parameters ``status`` and ``category``, then returns matching widgets.
-  it is used to populate respective containers and headers within the dashboard with the desired widgets.

**4. Dashboard is populated based on userId (this will be eventually based on logged in user)**

**5. Added tabs so other sections can be added by adding tab content. Notes page used as an example.**
- Additioanl pages should be ideally added as components to keep things simple and readable.

image

## Related Issue
#29 

## How Can This Be Tested?
`` npm run dev `` and go to ``/dashboard`` 

## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (if applicable)
- [x] My changes generate no new warnings, including `` npm run build ``
- [ ] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

File: src/components/Dashboard/Dashboard.tsx [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+	title: string;
+	children: React.ReactNode;
+}
+
+interface WidgetProps {
+	category: string;
+	[key: string]: any;
+}
+
+interface DahsboardProps {
```
Comment (halimahexe) [inline src/components/Dashboard/Dashboard.tsx:23]: Please correct spelling to `DashboardProps`

File: src/components/Dashboard/Dashboard.tsx [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+						category: child.props.category,
+					});
+				}
+				return child;
+			})}
+		</React.Fragment>
+	);
+}
+
+export function DashboardLayout({ userId }: DahsboardProps) {
```
Comment (halimahexe) [inline src/components/Dashboard/Dashboard.tsx:133]: And correct spelling of interface to `DashboardProps` here too, please

File: src/components/Dashboard/Dashboard.tsx [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+	const goalsData = useFetch<GoalsData>(`${baseUrl}/goals/user/${userId}`)
```
Comment (halimahexe) [inline src/components/Dashboard/Dashboard.tsx:134]: Please could you update the README to explain the baseUrl should be the same in both development and production envs?

File: src/components/Dashboard/Dashboard.tsx [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+		.data as GoalsData;
+
+	if (!goalsData) {
+		return (
+			<div className='flex items-center justify-center min-h-screen'>
+				<Loader2 className='w-12 h-12 animate-spin text-lilac' />
```
Comment (halimahexe) [inline src/components/Dashboard/Dashboard.tsx:140]: this is very cool! Could we console.log that there's an error, so devs can figure out what's wrong?

**Review by halimahexe — CHANGES_REQUESTED** [review]
Some really minor things I've spotted. Please could you fix and then I'll merge? Thank you!!

File: src/components/Dashboard/Dashboard.tsx (hunk 4 above)
Comment (oskarprzybylski23) [reply src/components/Dashboard/Dashboard.tsx:148]: I've changed this to handle error using in-built handling from useFetch function
```
	const [isLoading, setIsLoading] = useState(true);
	const fetchGoals = useFetch<GoalsData>(`${baseUrl}/goals/user/${userId}`);

	const goalsData = fetchGoals.data as GoalsData;

	useEffect(() => {
		if (fetchGoals.data || fetchGoals.error) {
			setIsLoading(false);
		}
	}, [goalsData, fetchGoals.error]);

	if (isLoading) {
		return (
			<div className='flex items-center justify-center min-h-screen'>
				<Loader2 className='w-12 h-12 animate-spin text-lilac' />
			</div>
		);
	}

	if (fetchGoals.error) {
		return (
			<div className='flex items-center justify-center min-h-screen'>
				<p className='text-red-500'>
					Error loading dashboard data. Please try again later.
				</p>
			</div>
		);
	}
	```

File: src/components/Dashboard/Dashboard.tsx (hunk 1 above)
Comment (oskarprzybylski23) [reply src/components/Dashboard/Dashboard.tsx]: done

File: src/components/Dashboard/Dashboard.tsx (hunk 3 above)
Comment (oskarprzybylski23) [reply src/components/Dashboard/Dashboard.tsx]: done

File: src/components/Dashboard/Dashboard.tsx (hunk 2 above)
Comment (oskarprzybylski23) [reply src/components/Dashboard/Dashboard.tsx]: done

File: src/components/Dashboard/Dashboard.tsx [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+	*/
+function addWidgets(
+	data: GoalsData,
+	status?:
+		| ('completed' | 'focused' | 'not_done' | null)[]
+		| ('completed' | 'focused' | 'not_done' | null),
+	category?: 'savings' | 'expenses' | 'investing' | 'charity'
+) {
+	if (!data || !data.categorizedGoals) return null;
+	console.log(data);
```
Comment (halimahexe) [inline src/components/Dashboard/Dashboard.tsx:66]: This is console logging the same thing multiple times. Do we still need this?

**Review by halimahexe — DISMISSED** [review]
I had one comment about a console.log but it's not breaking anything. Everything else looks good to me!

**Review by halimahexe — APPROVED** [review]
TY LGTM :)

**Pull request merged**